markdown/htmlparser.py

```diff
@@ -0,0 +1,187 @@
+            kwargs['convert_charrefs'] = False
+        # This calls self.reset
+        super().__init__(*args, **kwargs)
+        self.md = md
+
+    def reset(self):
+        """Reset this instance.  Loses all unprocessed data."""
+        self.inraw = False
+        self.stack = []  # When inraw==True, stack contains a list of tags
+        self._cache = []
+        self.cleandoc = []
+        super().reset()
+
+    def close(self):
+        """Handle any buffered data."""
+        super().close()
+        # Handle any unclosed tags.
+        if len(self._cache):
+            self.cleandoc.append(self.md.htmlStash.store(''.join(self._cache)))
+            self._cache = []
+
+    @property
+    def line_offset(self):
+        """Returns char index in self.rawdata for the start of the current line. """
+        if self.lineno > 1:
+            return re.match(r'([^\n]*\n){{{}}}'.format(self.lineno-1), self.rawdata).end()
+        return 0
+
+    def at_line_start(self):
+        """
+        Returns True if current position is at start of line.
+
+        Allows for up to three blank spaces at start of line.
+        """
+        if self.offset == 0:
+            return True
+        if self.offset > 3:
+            return False
+        # Confirm up to first 3 chars are whitespace
+        return self.rawdata[self.line_offset:self.line_offset + self.offset].strip() == ''
+
+    def get_endtag_text(self, tag):
+        """
+        Returns the text of the end tag.
+
+        If it fails to extract the actual text from the raw data, it builds a closing tag with `tag`.
+        """
+        # Attempt to extract actual tag from raw source text
+        start = self.line_offset + self.offset
+        m = htmlparser.endendtag.search(self.rawdata, start)
+        if m:
+            return self.rawdata[start:m.end()]
+        else:
+            # Failed to extract from raw data. Assume well formed and lowercase.
+            return '</{}>'.format(tag)
+
+    def handle_starttag(self, tag, attrs):
+        attrs = dict(attrs)
+
+        if self.at_line_start() and self.md.is_block_level(tag) and not self.inraw:
+            # Started a new raw block. Prepare stack.
+            self.inraw = True
+            self.cleandoc.append('\n')
+
+        text = self.get_starttag_text()
+        if self.inraw:
+            self.stack.append(tag)
+            self._cache.append(text)
+        else:
+            self.cleandoc.append(text)
+
+    def handle_endtag(self, tag):
+        text = self.get_endtag_text(tag)
+
+        if self.inraw:
+            self._cache.append(text)
+            if tag in self.stack:
+                # Remove tag from stack
+                while self.stack:
+                    if self.stack.pop() == tag:
+                        break
+            if len(self.stack) == 0:
+                # End of raw block.
+                if self.rawdata[self.line_offset + self.offset + len(text):].startswith('\n\n'):
```

Comment:
I think we need to discuss this. All this does is get the existing legacy tests to pass. However, it leaves a lot of unaddressed edge cases. To summarize, without this, All blank lines before and after a raw HTML block would be removed in the final output. And that is all fine as the extra white space is insignificant and does not affect a browser's rendering.

However, the reference implementation retains the blank lines, as did Python-Markdown in the past. Therefore, our old tests all include the blank lines. This change was the last to get all the old tests to pass.

That said, the behavior is inconsistent in practice. The old tests are simply incomplete in that they only include the common cases, but not the edge cases. For example, there could be spaces before or between the two newlines immediately following the closing tag of the HTML block. In those cases, the blank line would not be preserved.

I see three ways forward:

1. Leave things as is. The inconsistency is okay at it is all insignificant white space anyway.
2. Revert the last change and update the tests to not include the blank lines in the expected output.
3. Fix this properly by only closing out a raw block at the first newline following the closing tag. This would require the 'end of block' code to go in the `handle_data` method rather than the `handle_endtag` method. Then, where appropriate, the data would be split with everything up to and including the first newline being added to the raw block, with any remaining data being outside of the raw block.

Note that option 3 would also change the behavior of a number of tests which are not only related to white space. For example, consider these tests:

https://github.com/Python-Markdown/markdown/blob/efa36c8a03d29783d930d595cf341f34a080aa7d/tests/test_syntax/blocks/test_html_blocks.py#L703-L722

https://github.com/Python-Markdown/markdown/blob/efa36c8a03d29783d930d595cf341f34a080aa7d/tests/test_syntax/blocks/test_html_blocks.py#L983-L994

https://github.com/Python-Markdown/markdown/blob/efa36c8a03d29783d930d595cf341f34a080aa7d/tests/test_syntax/blocks/test_html_blocks.py#L1064-L1075

https://github.com/Python-Markdown/markdown/blob/efa36c8a03d29783d930d595cf341f34a080aa7d/tests/test_syntax/blocks/test_html_blocks.py#L1101-L1111

Implementing option 3 would bring all of those inline with the reference implementation.

The weird one is this:

https://github.com/Python-Markdown/markdown/blob/efa36c8a03d29783d930d595cf341f34a080aa7d/tests/test_syntax/blocks/test_html_blocks.py#L135-L146

Option 3 would restore Python-Markdown's behavior, but would not be inline with the reference implementation. Although, I think it would be reasonable to conclude that this is a bug in the reference implementation.

Personally, I'm fine with any of the options (preferring option 1 as it requires no work). I like option 2 because it is what I would have expected had I not checked the behavior of the reference implementation. Of course, option 3 actually most closely matches the reference implementation.

I will point out that I'm not all that concerned about changes in behavior here as they all involve content after the closing tag on the same line. I contend that users should not expect that to be allowed and any reliance on that within their documents is a hack.

Any feedback?

Reply:
I took a stab at option 3 and things started to get weird. So I took a less extreme approach. Mostly things stay as-is, except for two cases. The last last test case mentioned in previous comment is now:

```python
    def test_multiple_raw_single_line(self):
        self.assertMarkdownRenders(
            '<p>*foo*</p><div>*bar*</div>',
            self.dedent(
                """
                <p>*foo*</p>
                <div>*bar*</div>
                """
            )
        )
```

Also, the first test case in previous comment is now:

```python
    def test_raw_comment_one_line_followed_by_html(self):
        self.assertMarkdownRenders(
            '<!-- *foo* --><p>*bar*</p>',
            self.dedent(
                """
                <!-- *foo* -->
                <p>*bar*</p>
                """
            )
        )
```

The other three maintain their behavior, which I think makes more sense anyway. It is certainly what users would expect if they formatted their documents that way.

Reply:
Basically, I added a special case for a raw block which does not start at the beginning of a line, but immediately follows the end of the closing tag of a raw block on the same line as that closing tag.